src/jdk.incubator.code/share/classes/jdk/incubator/code/type/impl/JavaTypeUtils.java

```diff
@@ -0,0 +1,706 @@
+                            break;
+                        } else if (l.token(1).kind == Tokens.TokenKind.IDENTIFIER) {
+                            if (l.token(2).kind == Tokens.TokenKind.LPAREN || l.token(2).kind == Tokens.TokenKind.COLON) {
+                                // this looks like the middle of a field/method reference -- stop consuming
+                                break;
+                            }
+                            l.nextToken(); // inner type, keep going
+                        }
+                    } else {
+                        // not an inner type
+                        break;
+                    }
+                }
+            }
+        }
+        while (l.token().kind == Tokens.TokenKind.LBRACKET) {
+            l.accept(Tokens.TokenKind.LBRACKET);
+            l.accept(Tokens.TokenKind.RBRACKET);
+            type = JavaTypeUtils.arrayType(type);
+        }
+        return type;
+    }
+
+    private static ExternalizedTypeElement parseTypeVariableRest(ExternalizedTypeElement owner, Lexer l) {
+        l.accept(Tokens.TokenKind.LT);
+        String name = l.token().name();
+        l.nextToken();
+        ExternalizedTypeElement bound = JavaType.J_L_OBJECT.externalize();
+        if (l.token().kind == Tokens.TokenKind.IDENTIFIER &&
+                l.token().name().equals("extends")) {
+            l.nextToken();
+            bound = parseExternalTypeString(l);
+        }
+        l.accept(Tokens.TokenKind.GT);
+        return JavaTypeUtils.typeVarType(name, owner, bound);
+    }
+
+    private static ExternalizedTypeElement parseTypeArgument(Lexer l) {
+        if (l.token().kind == Tokens.TokenKind.QUES) {
+            // wildcard
+            l.nextToken();
+            ExternalizedTypeElement bound = JavaType.J_L_OBJECT.externalize();
+            WildcardType.BoundKind bk = BoundKind.EXTENDS;
+            if (l.token().kind == Tokens.TokenKind.IDENTIFIER) {
+                bk = switch (l.token().name()) {
+                    case "extends" -> BoundKind.EXTENDS;
+                    case "super" -> BoundKind.SUPER;
+                    default -> throw l.unexpected();
+                };
+                l.nextToken();
+                bound = parseExternalTypeString(l);
+            }
+            return JavaTypeUtils.wildcardType(bk, bound);
+        } else {
+            return parseExternalTypeString(l);
+        }
+    }
+
+    private static List<ExternalizedTypeElement> parseParameterTypes(Lexer l) {
+        List<ExternalizedTypeElement> ptypes = new ArrayList<>();
+        l.accept(Tokens.TokenKind.LPAREN);
+        if (l.token().kind != Tokens.TokenKind.RPAREN) {
+            ptypes.add(parseExternalTypeString(l));
+            while (l.acceptIf(Tokens.TokenKind.COMMA)) {
+                ptypes.add(parseExternalTypeString(l));
+            }
+        }
+        l.accept(Tokens.TokenKind.RPAREN);
+        return ptypes;
+    }
+
+    //    JavaRef:
```

Comment:
Note: the grammar of ref and types is partially overlapping -- for this reason it's slightly easier to keep the two parsing methods separate. If we wanted to unify more, we could require all refs to be prefixed with a `&` which would simplify things a bit (and also be compatible with what we do to disambiguate between class and method type-variable owner).

Reply:
Also, note: the grammar for method ref changed, as this now requires the type separator `:` to be used between the params and the return type (in a way that is more similar to what's done for field refs). If we don't want to have this incompatibility we can roll back.